File: python/src/wdl_model/model/processors/wdl_expression_processor.py
"""Visitor-style processor contract for WDL expression trees."""

from __future__ import annotations

from abc import ABC, abstractmethod

from wdl_model.model.expressions import (
    WdlArrayLiteral,
    WdlBinaryOperation,
    WdlBooleanLiteral,
    WdlExpression,
    WdlFloatLiteral,
    WdlFunctionCallOperation,
    WdlIndexAccessOperation,
    WdlIntLiteral,
    WdlMapEntry,
    WdlMapLiteral,
    WdlMemberAccessOperation,
    WdlNullLiteral,
    WdlObjectEntry,
    WdlObjectLiteral,
    WdlPairLiteral,
    WdlStringComponent,
    WdlStringEscape,
    WdlStringLiteral,
    WdlStringPlaceholder,
    WdlStringPlaceholderOption,
    WdlStringText,
    WdlStringToken,
    WdlStructEntry,
    WdlStructLiteral,
    WdlTernaryOperation,
    WdlUnaryOperation,
    WdlVariable,
)


class WdlExpressionProcessor(ABC):
    """Processor contract for walking expression trees."""

    @abstractmethod
    def processExpression(self, expression: WdlExpression | None) -> None: ...

    def dispatchByComponentType(self, expression: WdlExpression | None) -> None:
        """Dispatch helper for expression nodes based on concrete type."""
        if expression is None:
            return
        if isinstance(expression, WdlBooleanLiteral):
            self.onBooleanLiteral(expression)
        elif isinstance(expression, WdlFloatLiteral):
            self.onFloatLiteral(expression)
        elif isinstance(expression, WdlIntLiteral):
            self.onIntLiteral(expression)
        elif isinstance(expression, WdlArrayLiteral):
            self.onArrayLiteral(expression)
        elif isinstance(expression, WdlMapLiteral):
            self.onMapLiteral(expression)
        elif isinstance(expression, WdlNullLiteral):
            self.onNullLiteral(expression)
        elif isinstance(expression, WdlObjectLiteral):
            self.onObjectLiteral(expression)
        elif isinstance(expression, WdlPairLiteral):
            self.onPairLiteral(expression)
        elif isinstance(expression, WdlStringLiteral):
            self.onStringLiteral(expression)
        elif isinstance(expression, WdlStructLiteral):
            self.onStructLiteral(expression)
        elif isinstance(expression, WdlVariable):
            self.onVariable(expression)
        elif isinstance(expression, WdlBinaryOperation):
            self.onBinaryOperation(expression)
        elif isinstance(expression, WdlFunctionCallOperation):
            self.onFunctionCallOperation(expression)
        elif isinstance(expression, WdlIndexAccessOperation):
            self.onIndexAccessOperation(expression)
        elif isinstance(expression, WdlMemberAccessOperation):
            self.onMemberAccessOperation(expression)
        elif isinstance(expression, WdlTernaryOperation):
            self.onTernaryOperation(expression)
        elif isinstance(expression, WdlUnaryOperation):
            self.onUnaryOperation(expression)
        else:
            raise RuntimeError(
                f"Unhandled expression type: {expression.__class__.__name__}"
            )

    def dispatchStringComponentByType(
        self,
        context: WdlStringLiteral,
        component: WdlStringComponent | None,
    ) -> None:
        """Dispatch helper for string literal components by concrete type."""
        if component is None:
            return
        if isinstance(component, WdlStringText):
            self.onStringText(context, component)
        elif isinstance(component, WdlStringEscape):
            self.onStringEscape(context, component)
        elif isinstance(component, WdlStringToken):
            self.onStringToken(context, component)
        elif isinstance(component, WdlStringPlaceholder):
            self.onStringPlaceholder(context, component)
        else:
            raise RuntimeError(
                f"Unhandled string component type: {component.__class__.__name__}"
            )

    def onEnterExpression(self, expression: WdlExpression) -> None:
        pass

    def onExitExpression(self, expression: WdlExpression) -> None:
        pass

    def onEnterStringComponent(
        self, context: WdlStringLiteral, component: WdlStringComponent
    ) -> None:
        pass

    def onExitStringComponent(
        self, context: WdlStringLiteral, component: WdlStringComponent
    ) -> None:
        pass

    def onBooleanLiteral(self, expression: WdlBooleanLiteral) -> None:
        pass

    def onFloatLiteral(self, expression: WdlFloatLiteral) -> None:
        pass

    def onIntLiteral(self, expression: WdlIntLiteral) -> None:
        pass

    def onNullLiteral(self, expression: WdlNullLiteral) -> None:
        pass

    def onVariable(self, expression: WdlVariable) -> None:
        pass

    def onArrayLiteral(self, expression: WdlArrayLiteral) -> None:
        pass

    def onMapLiteral(self, expression: WdlMapLiteral) -> None:
        pass

    def onMapEntry(self, context: WdlMapLiteral, entry: WdlMapEntry) -> None:
        pass

    def onObjectLiteral(self, expression: WdlObjectLiteral) -> None:
        pass

    def onObjectEntry(self, context: WdlObjectLiteral, entry: WdlObjectEntry) -> None:
        pass

    def onPairLiteral(self, expression: WdlPairLiteral) -> None:
        pass

    def onStringLiteral(self, expression: WdlStringLiteral) -> None:
        pass

    def onStringText(self, context: WdlStringLiteral, text: WdlStringText) -> None:
        pass

    def onStringEscape(
        self, context: WdlStringLiteral, escape: WdlStringEscape
    ) -> None:
        pass

    def onStringToken(self, context: WdlStringLiteral, token: WdlStringToken) -> None:
        pass

    def onStringPlaceholder(
        self, context: WdlStringLiteral, placeholder: WdlStringPlaceholder
    ) -> None:
        pass

    def onStringPlaceholderOption(
        self,
        context: WdlStringLiteral,
        placeholder: WdlStringPlaceholder,
        option: WdlStringPlaceholderOption,
    ) -> None:
        pass

    def onStructLiteral(self, expression: WdlStructLiteral) -> None:
        pass

    def onStructEntry(self, context: WdlStructLiteral, entry: WdlStructEntry) -> None:
        pass

    def onBinaryOperation(self, expression: WdlBinaryOperation) -> None:
        pass

    def onFunctionCallOperation(self, expression: WdlFunctionCallOperation) -> None:
        pass

    def onIndexAccessOperation(self, expression: WdlIndexAccessOperation) -> None:
        pass

    def onMemberAccessOperation(self, expression: WdlMemberAccessOperation) -> None:
        pass

    def onTernaryOperation(self, expression: WdlTernaryOperation) -> None:
        pass

    def onUnaryOperation(self, expression: WdlUnaryOperation) -> None:
        pass

File: java/src/main/java/com/myriad/wdl/model/processors/WdlExpressionProcessor.java
package com.myriad.wdl.model.processors;

import com.myriad.wdl.model.expressions.WdlArrayLiteral;
import com.myriad.wdl.model.expressions.WdlBinaryOperation;
import com.myriad.wdl.model.expressions.WdlBooleanLiteral;
import com.myriad.wdl.model.expressions.WdlExpression;
import com.myriad.wdl.model.expressions.WdlFloatLiteral;
import com.myriad.wdl.model.expressions.WdlFunctionCallOperation;
import com.myriad.wdl.model.expressions.WdlIndexAccessOperation;
import com.myriad.wdl.model.expressions.WdlIntLiteral;
import com.myriad.wdl.model.expressions.WdlMapLiteral;
import com.myriad.wdl.model.expressions.WdlMapLiteral.WdlMapEntry;
import com.myriad.wdl.model.expressions.WdlMemberAccessOperation;
import com.myriad.wdl.model.expressions.WdlNullLiteral;
import com.myriad.wdl.model.expressions.WdlObjectLiteral;
import com.myriad.wdl.model.expressions.WdlObjectLiteral.WdlObjectEntry;
import com.myriad.wdl.model.expressions.WdlPairLiteral;
import com.myriad.wdl.model.expressions.WdlStringLiteral;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringComponent;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringEscape;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringPlaceholder;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringPlaceholderOption;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringText;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringToken;
import com.myriad.wdl.model.expressions.WdlStructLiteral;
import com.myriad.wdl.model.expressions.WdlStructLiteral.WdlStructEntry;
import com.myriad.wdl.model.expressions.WdlTernaryOperation;
import com.myriad.wdl.model.expressions.WdlUnaryOperation;
import com.myriad.wdl.model.expressions.WdlVariable;

/**
 * Processor contract for walking expression trees.
 *
 * <p>The traversal strategy is intentionally unspecified so implementations can do depth-first,
 * breadth-first, or custom traversals.
 *
 * <p>{@link #dispatchByComponentType(WdlExpression)} and
 * {@link #dispatchStringComponentByType(WdlStringLiteral, WdlStringComponent)} are provided as
 * reusable component-type dispatch helpers to avoid {@code instanceof} chains in implementers.
 */
public interface WdlExpressionProcessor {
  /**
   * Process an expression tree rooted at {@code expression}.
   *
   * <p>Implementers own traversal order and recursion/queue policy.
   */
  void processExpression(WdlExpression expression);

  /** Dispatch helper for expression nodes based on {@code componentType()}. */
  default void dispatchByComponentType(WdlExpression expression) {
    if (expression == null) {
      return;
    }

    switch (expression.componentType()) {
      case BOOL_LIT:
        onBooleanLiteral((WdlBooleanLiteral) expression);
        break;
      case FLOAT_LIT:
        onFloatLiteral((WdlFloatLiteral) expression);
        break;
      case INT_LIT:
        onIntLiteral((WdlIntLiteral) expression);
        break;
      case ARRAY_LIT:
        onArrayLiteral((WdlArrayLiteral) expression);
        break;
      case MAP_LIT:
        onMapLiteral((WdlMapLiteral) expression);
        break;
      case NULL_LIT:
        onNullLiteral((WdlNullLiteral) expression);
        break;
      case OBJ_LIT:
        onObjectLiteral((WdlObjectLiteral) expression);
        break;
      case PAIR_LIT:
        onPairLiteral((WdlPairLiteral) expression);
        break;
      case STR_LIT:
        onStringLiteral((WdlStringLiteral) expression);
        break;
      case STRUCT_LIT:
        onStructLiteral((WdlStructLiteral) expression);
        break;
      case VARIABLE:
        onVariable((WdlVariable) expression);
        break;
      case BINARY_OP:
        onBinaryOperation((WdlBinaryOperation) expression);
        break;
      case FUNC_OP:
        onFunctionCallOperation((WdlFunctionCallOperation) expression);
        break;
      case IDX_OP:
        onIndexAccessOperation((WdlIndexAccessOperation) expression);
        break;
      case MEMBER_OP:
        onMemberAccessOperation((WdlMemberAccessOperation) expression);
        break;
      case TERNARY_OP:
        onTernaryOperation((WdlTernaryOperation) expression);
        break;
      case UNARY_OP:
        onUnaryOperation((WdlUnaryOperation) expression);
        break;
      default:
        throw new IllegalStateException(
            "Unhandled expression component type: " + expression.componentType());
    }
  }

  /** Dispatch helper for string components based on {@code componentType()}. */
  default void dispatchStringComponentByType(
      WdlStringLiteral context, WdlStringComponent component) {
    if (component == null) {
      return;
    }

    switch (component.componentType()) {
      case TEXT:
        onStringText(context, (WdlStringText) component);
        break;
      case ESC:
        onStringEscape(context, (WdlStringEscape) component);
        break;
      case SPECIAL:
        onStringToken(context, (WdlStringToken) component);
        break;
      case PLACEHOLDER:
        onStringPlaceholder(context, (WdlStringPlaceholder) component);
        break;
      default:
        throw new IllegalStateException(
            "Unhandled string component type: " + component.componentType());
    }
  }

  default void onEnterExpression(WdlExpression expression) {}

  default void onExitExpression(WdlExpression expression) {}

  default void onEnterStringComponent(WdlStringLiteral context, WdlStringComponent component) {}

  default void onExitStringComponent(WdlStringLiteral context, WdlStringComponent component) {}

  default void onBooleanLiteral(WdlBooleanLiteral expression) {}

  default void onFloatLiteral(WdlFloatLiteral expression) {}

  default void onIntLiteral(WdlIntLiteral expression) {}

  default void onNullLiteral(WdlNullLiteral expression) {}

  default void onVariable(WdlVariable expression) {}

  default void onArrayLiteral(WdlArrayLiteral expression) {}

  default void onMapLiteral(WdlMapLiteral expression) {}

  default void onMapEntry(WdlMapLiteral context, WdlMapEntry entry) {}

  default void onObjectLiteral(WdlObjectLiteral expression) {}

  default void onObjectEntry(WdlObjectLiteral context, WdlObjectEntry entry) {}

  default void onPairLiteral(WdlPairLiteral expression) {}

  default void onStringLiteral(WdlStringLiteral expression) {}

  default void onStringText(WdlStringLiteral context, WdlStringText text) {}

  default void onStringEscape(WdlStringLiteral context, WdlStringEscape escape) {}

  default void onStringToken(WdlStringLiteral context, WdlStringToken token) {}

  default void onStringPlaceholder(WdlStringLiteral context, WdlStringPlaceholder placeholder) {}

  default void onStringPlaceholderOption(
      WdlStringLiteral context,
      WdlStringPlaceholder placeholder,
      WdlStringPlaceholderOption option) {}

  default void onStructLiteral(WdlStructLiteral expression) {}

  default void onStructEntry(WdlStructLiteral context, WdlStructEntry entry) {}

  default void onBinaryOperation(WdlBinaryOperation expression) {}

  default void onFunctionCallOperation(WdlFunctionCallOperation expression) {}

  default void onIndexAccessOperation(WdlIndexAccessOperation expression) {}

  default void onMemberAccessOperation(WdlMemberAccessOperation expression) {}

  default void onTernaryOperation(WdlTernaryOperation expression) {}

  default void onUnaryOperation(WdlUnaryOperation expression) {}
}
